R/layer_population_scaling.R

```diff
@@ -0,0 +1,149 @@
+#'   dplyr::filter(time_value > "2021-11-01", geo_value %in% c("ca", "ny")) %>%
+#'   dplyr::select(geo_value, time_value, cases)
+#'
+#' pop_data = data.frame(states = c("ca", "ny"),
+#'                       value = c(20000, 30000))
+#'
+#' r <- epi_recipe(jhu) %>%
+#'   step_population_scaling(df = pop_data,
+#'                           df_pop_col = "value",
+#'                           by = c("geo_value" = "states"),
+#'                           cases, suffix = "_scaled") %>%
+#'   step_epi_lag(cases_scaled, lag = c(7, 14)) %>%
+#'   step_epi_ahead(cases_scaled, ahead = 7, role = "outcome") %>%
+#'   step_epi_naomit()
+#'
+#' f <- frosting() %>%
+#'   layer_predict() %>%
+#'   layer_threshold(.pred) %>%
+#'   layer_naomit(.pred) %>%
+#'   layer_population_scaling(.pred, df = pop_data,
+#'                            by =  c("geo_value" = "states"),
+#'                            df_pop_col = "value")
+#'
+#' wf <- epi_workflow(r,
+#'                    parsnip::linear_reg()) %>%
+#'   parsnip::fit(jhu) %>%
+#'   add_frosting(f)
+#'
+#' latest <- get_test_data(recipe = r,
+#'                         x = epiprocess::jhu_csse_daily_subset %>%
+#'                           dplyr::filter(time_value > "2021-11-01",
+#'                                         geo_value %in% c("ca", "ny")) %>%
+#'                           dplyr::select(geo_value, time_value, cases))
+#'
+#'
+#' predict(wf, latest)
+layer_population_scaling <- function(frosting,
+                           ...,
+                           df,
+                           by = NULL,
+                           df_pop_col,
+                           create_new = TRUE,
+                           suffix = "_original",
+                           .flag = TRUE,
+                           id = rand_id("population_scaling")) {
+
+  add_layer(
+    frosting,
+    layer_population_scaling_new(
+      df = df,
+      by = by,
+      df_pop_col = df_pop_col,
+      terms = dplyr::enquos(...),
+      create_new = create_new,
+      suffix = suffix,
+      id = id
+    ),
+    flag = .flag
+  )
+}
+
+layer_population_scaling_new <-
+  function(df, by, df_pop_col, terms, create_new, suffix, id) {
+  layer("population_scaling",
+        df = df,
+        by = by,
+        df_pop_col = df_pop_col,
+        terms = terms,
+        create_new = create_new,
+        suffix = suffix,
+        id = id)
+}
+
+#' @export
+slather.layer_population_scaling <-
+  function(object, components, the_fit, the_recipe, ...) {
+    stopifnot("Only one population column allowed for scaling" =
+                length(object$df_pop_col) == 1)
+
+    t <- try(dplyr::left_join(components$predictions, object$df,
+                              by= tolower(object$by)),
+             silent = TRUE)
+    if (any(grepl("Join columns must be present in data", unlist(t)))){
+      stop("columns in `by` selectors of `layer_population_scaling` must be present in data and match")}
+
```

Comment:
So in the `bake` function, I intended to use the `try()` function on the left join to see if the `by` selector functions properly, and stop/throw an error accordingly. The actual left join is then performed later in the `bake` function. I used this method  instead of the usual `ifelse` check because of how the `by` selector are written. Sometimes it's written as `c("a" = "b")` or some other formats which is hard to assess. 

What do you think of the this check? Does is match what you're thinking of?